### PR TITLE
Loginregister implementation

### PR DESCRIPTION
**Pull request by JannisAziz**
+ Sign in, sign up & email confirmation implementation

+ Game search, display & save implementation (should have its own branch, I know, sorry..)

Look at the commits messages before reviewing, makes it more obvious what changed.

### CODE REVIEW

**Review by RedStylzZ — APPROVED** [review]
LGTM

File: backend/src/main/java/de/jannisaziz/backend/game/RAWGGameRepository.java [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
-import de.jannisaziz.backend.models.Game;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Repository;
 import org.springframework.web.reactive.function.client.WebClient;
 import reactor.core.publisher.Mono;
-import utils.RAWGResponseObject;
+import de.jannisaziz.backend.game.utils.RAWGResponseObject;
 
 import java.util.List;
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/game/RAWGGameRepository.java:11]:
```suggestion
public class RAWGGameApiService
```

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+@AllArgsConstructor
+@RestController
+@RequestMapping("/api/auth")
+public class LoginController {
+
+    private final LoginService loginService;
+
+    @PostMapping("signIn")
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:18]:
```suggestion
    @PostMapping("signin")
```

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    public UserDTO signIn(@RequestBody LoginRequest request) throws ResponseStatusException {
+        try {
+            return loginService.signIn(request).asDTO();
+        } catch (IllegalStateException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    @PostMapping("signUp")
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:27]:
```suggestion
    @PostMapping("signup")
```

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    public String signUp(@RequestBody LoginRequest request) throws ResponseStatusException {
+        try {
+            return loginService.signUp(request); // Returns "Confirmation email sent" on success
+        } catch (IllegalStateException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    @PostMapping("resendConfirmation={email}")
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:36]:
```suggestion
    @PostMapping("resendconfirmation={email}")
```

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    public String resendConfirmationEmail(@PathVariable String email) throws ResponseStatusException {
+        try {
+            return loginService.resendConfirmationEmail(email);
+        } catch (IllegalArgumentException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    @GetMapping("confirm={confirmationToken}")
+    public String confirm(HttpServletResponse response, @PathVariable String confirmationToken) throws ResponseStatusException {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:46]: Consider using RequestParams here. Url syntax: [...]?confirm="token"

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+        response.setHeader("Location", "http://localhost:3000/");
+        response.setStatus(302);
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:49]: Why are you manipulating the request here? 🤔

File: backend/src/main/java/de/jannisaziz/backend/security/email/EmailService.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        Properties props = new Properties();
+        props.put("mail.smtp.host", "smtp.gmail.com");
+        props.put("mail.smtp.port", "587");
+        props.put("mail.smtp.auth", "true");
+        props.put("mail.smtp.starttls.enable", "true");
+
+        Session session = Session.getInstance(props, new Authenticator() {
+            @Override
+            protected PasswordAuthentication getPasswordAuthentication() {
+                return new PasswordAuthentication("capstoneludomemoria", "ludo-memoria");
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/email/EmailService.java:26]: Dont hardcode credentials

File: backend/src/main/java/de/jannisaziz/backend/security/LoginService.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            if (user.getPassword().equals(encodedRequestPassword)) return user;
+            else throw new IllegalArgumentException("Wrong Password");
+        } catch (UsernameNotFoundException | IllegalArgumentException e) {
+            throw new IllegalStateException(e.getMessage());
+        }
+    }
+
+    public String signUp(LoginRequest request) throws IllegalStateException {
+        try {
+            String confirmationToken = UUID.randomUUID().toString();
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginService.java:36]: I would think about using a jwt as confirmation token, then you wouldn't need to save it to the db

File: backend/src/main/java/de/jannisaziz/backend/user/UserController.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+@AllArgsConstructor
+@RestController
+@RequestMapping("/api/users")
+public class UserController {
+
+    private final UserService service;
+
+    @GetMapping("/id/{userId}")
+    public User getUserById(@PathVariable String userId) throws ResponseStatusException {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/user/UserController.java:17]: Dont return the whole user including password and other private data like email-adress

File: backend/src/main/java/de/jannisaziz/backend/user/UserController.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        try {
+            return service.getUserById(userId);
+        } catch (UsernameNotFoundException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    @GetMapping("/name/{username}")
+    public User getUserByUsername(@PathVariable String username) throws ResponseStatusException {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/user/UserController.java:26]: Dont return the whole user including password and other private data like email-adress

File: backend/src/main/java/de/jannisaziz/backend/user/UserController.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        try {
+            return service.findUser(username, null);
+        } catch (IllegalArgumentException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    @PatchMapping("/update")
+    public String updateUser(@RequestBody UserDTO userDTO) throws ResponseStatusException {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/user/UserController.java:35]: Why not return the updated user object?

File: backend/src/main/java/de/jannisaziz/backend/user/UserDTO.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+import de.jannisaziz.backend.game.Review;
+import de.jannisaziz.backend.game.SavedGame;
+import lombok.*;
+import org.springframework.data.annotation.Id;
+import java.util.List;
+
+@Builder
+@Getter
+@Setter
+public class UserDTO {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/user/UserDTO.java:13]: What is this dto used for? Has a lot of private data, why does the user need to see his or others oauthConnections?

File: backend/src/main/java/de/jannisaziz/backend/user/UserService.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public String confirmUser(String email) throws UsernameNotFoundException {
+        User user = userRepository.findByUsernameOrEmail(null, email)
+                .orElseThrow(() -> USR_NF_EX(email));
+
+        user.setEnabled(true);
+        userRepository.save(user);
+
+        return "User confirmed";
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/user/UserService.java:73]: Is this return necessary? I would maybe return true? 🤔

File: frontend/src/components/CatalogPage.tsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+import SearchField from "./SearchField";
+import {Game} from "../games/Game";
+import {findGamesByName} from "../games/GameService";
+
+export default function CatalogPage() {
+
+    const [games, setGames] = useState<Game[]>([])
+
+    const onSearch = (input: string) => {
+        console.log(`Search: ${input}`)
```
Comment (Flooooooooooorian) [inline frontend/src/components/CatalogPage.tsx:12]:
```suggestion
```

File: frontend/src/games/GameCard.tsx [hunk 15]
```diff
@@ -0,0 +1,26 @@
+import {Game} from "./Game";
+import {MouseEventHandler} from "react";
+import {saveGameToLibrary} from "../users/UserService";
+
+export default function GameCard({game}: {game: Game}) {
+
+    const onCardPressed: MouseEventHandler<HTMLButtonElement> = (e) => {
+        e.preventDefault()
+
+        console.log(`Open game page for ${game.name}`)
```
Comment (Flooooooooooorian) [inline frontend/src/games/GameCard.tsx:10]:
```suggestion
```

File: frontend/src/games/GameCard.tsx [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    const onSaveGamePressed: MouseEventHandler<HTMLButtonElement> = (e) => {
+        e.preventDefault()
```
Comment (Flooooooooooorian) [inline frontend/src/games/GameCard.tsx:14]: event.preventDefault() is only necessary when a form is submitted.
```suggestion
```

File: frontend/src/games/GameCard.tsx [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import {Game} from "./Game";
+import {MouseEventHandler} from "react";
+import {saveGameToLibrary} from "../users/UserService";
+
+export default function GameCard({game}: {game: Game}) {
+
+    const onCardPressed: MouseEventHandler<HTMLButtonElement> = (e) => {
+        e.preventDefault()
```
Comment (Flooooooooooorian) [inline frontend/src/games/GameCard.tsx:8]: event.preventDefault() is only necessary when a form is submitted.
```suggestion
```

File: backend/src/main/java/de/jannisaziz/backend/game/GameService.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+public class GameService {
+
+    private static IllegalArgumentException NO_GAMES_FOUND_EX(String identifier) {
+        return new IllegalArgumentException("No games found: " + identifier);
+    }
+
+    private final GameRepository repository;
+    private final RAWGGameRepository rawgRepository;
+
+    private boolean isOnline = true;
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/game/GameService.java:19]: What is this for?

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    public UserDTO signIn(@RequestBody LoginRequest request) throws ResponseStatusException {
+        try {
+            return loginService.signIn(request).asDTO();
+        } catch (IllegalStateException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    @PostMapping("signUp")
+    public String signUp(@RequestBody LoginRequest request) throws ResponseStatusException {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:28]: Maybe return just a bool or a singupRresponseDto with emailConfirmation: true and the email: <user.email>?

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+        try {
+            return loginService.signUp(request); // Returns "Confirmation email sent" on success
+        } catch (IllegalStateException e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, e.getMessage());
+        }
+    }
+
+    @PostMapping("resendConfirmation={email}")
+    public String resendConfirmationEmail(@PathVariable String email) throws ResponseStatusException {
```
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:37]: Maybe return just a bool or a singupRresponseDto with emailConfirmationSend: true and the email: <user.email>?

File: backend/src/main/java/de/jannisaziz/backend/security/LoginController.java (hunk 5 above)
Comment (Flooooooooooorian) [inline backend/src/main/java/de/jannisaziz/backend/security/LoginController.java:46]: Maybe return just a bool or maybe the token so the user is loggedin?

**Review by Flooooooooooorian — APPROVED** [review]
Im not sure what User.username and User.email is? You are using it kinda confusing 🤔

**Pull request closed without merging**